### DISCUSSION TITLE
cannot resolve these dependencies

### DISCUSSION THREAD

**Discussion by oldRabbitForz** [internetarchive/heritrix3#520] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/38576501/138016636-6e12d82e-2b8c-46df-910a-cd3c5a3b96c8.png)

**Comment by oldRabbitForz** [comment]
<dependency>
            <groupId>org.archive.heritrix</groupId>
            <artifactId>heritrix-commons</artifactId>
            <version>3.4.0-20210923</version>
        </dependency>
        <dependency>
            <groupId>org.archive.heritrix</groupId>
            <artifactId>heritrix-modules</artifactId>
            <version>3.4.0-20210923</version>
        </dependency>
        <dependency>
            <groupId>org.archive.heritrix</groupId>
            <artifactId>heritrix-engine</artifactId>
            <version>3.4.0-20210923</version>
        </dependency>

**Comment by oldRabbitForz** [comment]
@ato

**Comment by ato** [comment]
I assume that's a screenshot from some IDE but some more context would be helpful. :-) 
What are you trying to do? What software are you using?

That looks as if the extra repositories that Heritrix requires are not being searched. com.sleepycat:je is in https://download.oracle.com/maven and the other three are in http://builds.archive.org/maven2/ (which recent versions of Maven refuse to access by default without a [workaround](https://github.com/internetarchive/heritrix3/pull/310#issuecomment-857543812))

**Comment by oldRabbitForz** [comment]
i'm using idea ide.  a program with Heritrix3 based on springboot.

**Comment by oldRabbitForz** [comment]
when i download these dependencies, the mistakes occured.so how  can i resolve these problems.thanks!

**Comment by oldRabbitForz** [comment]
@ato

**Comment by ato** [comment, marked as answer]
If you're including it as a dependency in a Maven project maybe try adding these repositories to your pom.xml file, although I would normally expect them to be included automatically. You may also need [Andy's ~/.m2/settings.xml workaround](https://github.com/internetarchive/heritrix3/pull/310#issuecomment-857543812).

```xml
    <repositories>
        <repository>
            <id>builds.archive.org,maven2</id>
            <url>http://builds.archive.org/maven2</url>
        </repository>
        <repository>
            <id>oracleReleases</id>
            <name>Oracle Released Java Packages</name>
            <url>https://download.oracle.com/maven</url>
        </repository>
    </repositories>
```

If you're using Gradle, I can't help as I've never used it.

However personally I don't recommend embedding Heritrix inside another Java application as it has a lot of dependencies that may cause conflicts and it also does some surprising things like globally seting the JVM's timezone to UTC. I recommend controlling it via the [REST API](https://heritrix.readthedocs.io/en/latest/api.html) if possible.

**Comment by oldRabbitForz** [comment]
thanks,when the crawl data come back，i want process them for other biz. if i use the rest api, i must build another data table to translate the data. i would be complicated。

**Comment by oldRabbitForz** [comment]
it would be complicated。

**Comment by oldRabbitForz** [comment]
@ato
